### PR TITLE
#160451496  Return JWT on Successful Registration

### PR DESCRIPTION
**Pull request by udochukwu**
**What does this PR do?**

Users should receive a JWT on Successful Registration

**Description of Task to be completed?**

- modify the user migrations and models to accept username

- install Jsonwebtoken and create auth middlewares for signing and verifying tokens

- modify the create method under userController to return a token on successful signup

-  write tests for the user signup 

- add a .eslintignore file to skip Auto-generated files 

**What are the relevant pivotal tracker stories?**

#160451496

### CODE REVIEW

**Review by onwuvic — COMMENTED** [review]
LTGM

**Comment by kampkelly** [comment]
LGTM

File: server/controllers/userController.js [hunk 1]
```diff
@@ -1,22 +1,38 @@
+import bcrypt from 'bcrypt';
 import Model from '../models';
+import Authenticator from '../middlewares/authenticator';
 
 const { User } = Model;
```
Comment (s-akinrele) [inline server/controllers/userController.js:5]: `import {User} from '../models';`

File: server/middlewares/authenticator.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import jwt from 'jsonwebtoken';
+import 'dotenv/config';
+
+export default {
+  generateToken(user) {
+    const token = jwt.sign({ user }, process.env.JWTKEY, { expiresIn: '10000s' });
```
Comment (s-akinrele) [inline server/middlewares/authenticator.js]: lets update this to 2 days

File: server/routes/api/user.js [hunk 3]
```diff
@@ -1,14 +1,13 @@
-import express from "express"
-import UserController  from '../../controllers/userController';
+import express from 'express';
+import UserController from '../../controllers/userController';
 
 const router = express.Router();
 
 /* GET home page. */
-router.get('/', (req, res, next) => res.json('welcome Author Haven'));
+router.get('/', (req, res) => res.json('welcome Author Haven'));
```
Comment (s-akinrele) [inline server/routes/api/user.js]: please put back the deleted code

**Review by s-akinrele — CHANGES_REQUESTED** [review]
Kindly add more tests and update the PR.

File: server/controllers/userController.js (hunk 1 above)
Comment (udochukwu) [reply server/controllers/userController.js:5]: I tried this solution it didn't work. I went to stack overflow and this is the solution  https://stackoverflow.com/questions/33524696/es6-destructuring-and-module-imports

File: server/controllers/userController.js [hunk 4]
```diff
@@ -1,22 +1,38 @@
...
-    User.create({
-      email,
-      password
-    })
-    .then((user) => res.status(201).json(user))
-    .catch((err) => res.status(500).json(err));
+    const { username, email, password } = req.body;
+    bcrypt.hash(password, saltRounds)
+      .then(hashedPassword => User.create({
+        username,
```
Comment (s-akinrele) [inline server/controllers/userController.js]: can this be on a single line ``` .then(hashedPassword => User.create({ username, email, password: hashedPassword })```

File: server/controllers/userController.js [hunk 5]
```diff
@@ -1,22 +1,38 @@
...
+        email,
+        password: hashedPassword,
+      })
+        .then((user) => {
+          const { userid } = user;
+          const token = generateToken({ userid });
+          res.status(201).json({
```
Comment (s-akinrele) [inline server/controllers/userController.js]: same applies here

File: server/middlewares/authenticator.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      expiresIn: process.env.TOKEN_EXPIRATION
+    });
+    return token;
+  },
+
+  checkToken(req, res, next) {
+    const token = req.body.token || req.headers['x-access-token'] || req.headers.bearer;
+    if (!token) {
+      res.status(403)
+        .json({
```
Comment (s-akinrele) [inline server/middlewares/authenticator.js]: can this be on a single line

File: server/middlewares/authenticator.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+          success: false,
+          message: 'Missing Token'
+        });
+    } else {
+      jwt.verify(token, process.env.JWTKEY, (err, decoded) => {
+        if (err) {
+          if (err.message.includes('signature')) {
+            res.status(403)
```
Comment (s-akinrele) [inline server/middlewares/authenticator.js]: no need for multiple lines

**Review by s-akinrele — COMMENTED** [review]
I had some comments, also I am yet to see the model test. @udochukwu

File: server/tests/user.test.js [hunk 8]
```diff
@@ -3,21 +3,61 @@ import chaiHttp from 'chai-http';
...
+      .set('x-access-token', token)
+      .end((err, res) => {
+        expect(res.body).to.be.an('object');
+        done();
+      });
+  });
+});
+
+describe('Access a protected route without any token', () => {
+  it('SShould throw an error 403', (done) => {
```
Comment (s-akinrele) [inline server/tests/user.test.js]: this should be `Should` @udochukwu

Comment (udochukwu) [reply server/tests/user.test.js]: I have corrected the error.

File: server/index.js [hunk 9]
```diff
@@ -1,23 +1,25 @@
+/* eslint-disable no-unused-vars */
 import fs from 'fs';
 import http from 'http';
 import path from 'path';
 import express from 'express';
 import bodyParser from 'body-parser';
 import cors from 'cors';
+/* eslint-disable no-unused-vars */
```
Comment (s-akinrele) [inline server/index.js]: Can you add this rule to eslint file

**Review by s-akinrele — APPROVED** [review]

**Pull request merged**